File: 04/challenge-04.ts
import { readFileSync } from 'node:fs'

function isReal (filename: string): boolean {
  const [id, checksum] = filename.split('-')
  const uniques = Array.from(id).filter((ch, index, arr) => arr.indexOf(ch) === arr.lastIndexOf(ch)).join('')
  return uniques === checksum
}

function main (): void {
  console.log('Challenge #4')
  const file = readFileSync('./04/in').toString()
  const lines = file.split('\n')
  lines.pop()
  const reals = lines.filter(isReal)
  console.log(`Real file #33 checksum: ${reals[32].split('-')[1]}`)
}

main()

File: 04/challenge-04.hs
import Data.List.Split (splitOn)


firstIndex :: String -> Char -> Int
firstIndex str c = do
  let (x:xs) = str
  if x == c then 0 else 1 + firstIndex xs c

lastIndexAux :: String -> Char -> Int -> Int
lastIndexAux str char index
  | index < 0 = -1
  | str !! index == char = index
  | otherwise = lastIndexAux str char (index - 1)

lastIndex :: String -> Char -> Int
lastIndex str c = lastIndexAux str c (length str - 1)

isUnique :: String -> Char -> Bool
isUnique str c = firstIndex str c == lastIndex str c

reduceToUnique :: String -> String
reduceToUnique str = [x | x <- str, isUnique str x]

getPathAndChecksum :: String -> (String, String)
getPathAndChecksum filename = (path, checksum)
  where
    [path, checksum] = splitOn "-" filename

isRealFile :: String -> Bool
isRealFile filename = checksum == reduceToUnique path
  where
    (path, checksum) = getPathAndChecksum filename

main :: IO()
main = do
  let filename = "./04/in"
  contents <- readFile filename
  let lines = init (splitOn "\n" contents)
  let realFiles = filter isRealFile lines
  print "Real File #33 checksum:"
  print $ snd (getPathAndChecksum (realFiles !! 32))
